File: algorithms/bfs/bfs.py
from collections import deque
from typing import List


def bfs(graph: List[List[int]], start: int) -> List[int]:
    """Return the BFS traversal order from start."""
    visited = [False] * len(graph)
    order = []
    queue = deque([start])
    visited[start] = True

    while queue:
        vertex = queue.popleft()
        order.append(vertex)

        for neighbor in graph[vertex]:
            if not visited[neighbor]:
                visited[neighbor] = True
                queue.append(neighbor)

    return order


if __name__ == "__main__":
    graph = [
        [1, 2],
        [0, 3, 4],
        [0, 5],
        [1],
        [1],
        [2],
    ]

    print("BFS traversal:", bfs(graph, 0))

File: algorithms/bfs/bfs.java
import java.util.*;

public class BFS {
    static List<Integer> bfs(List<List<Integer>> graph, int start) {
        boolean[] visited = new boolean[graph.size()];
        List<Integer> order = new ArrayList<>();
        Queue<Integer> queue = new ArrayDeque<>();

        visited[start] = true;
        queue.offer(start);

        while (!queue.isEmpty()) {
            int vertex = queue.poll();
            order.add(vertex);

            for (int neighbor : graph.get(vertex)) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    queue.offer(neighbor);
                }
            }
        }

        return order;
    }

    public static void main(String[] args) {
        List<List<Integer>> graph = new ArrayList<>();
        graph.add(Arrays.asList(1, 2));
        graph.add(Arrays.asList(0, 3, 4));
        graph.add(Arrays.asList(0, 5));
        graph.add(Collections.singletonList(1));
        graph.add(Collections.singletonList(1));
        graph.add(Collections.singletonList(2));

        System.out.println("BFS traversal: " + bfs(graph, 0));
    }
}
